### PR TITLE
Detection: panoptic quality

### LINKED ISSUE #50: [Metrics] Panoptic Quality
## 🚀 Feature
Implement Panoptic Quality

### PR DESCRIPTION
**Pull request by niberger**
## What does this PR do?

Add the panoptic quality metric, as defined in the [reference paper](https://arxiv.org/abs/1801.00868).
This implementation differ from the [reference implementation](https://github.com/cocodataset/panopticapi/blob/master/panopticapi/evaluation.py) of the COCO dataset on the following points:
- The categories have to be declared at metric initialization.
- Any undefined category will be considered as an unlabelled (void) region.
- The input format for panoptic images are `Tensor` of shape `(height, width, 2)`, where the first number of each pixel is the category id and the second number is the instance id.
- No support for group of instances (crowds in COCO).

Fixes #50
closes #1435

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/references/modules.rst [hunk 1]
```diff
@@ -462,6 +462,12 @@ MeanAveragePrecision
 .. autoclass:: torchmetrics.detection.mean_ap.MeanAveragePrecision
     :noindex:
 
+PanopticQuality
+~~~~~~~~~~~~~~~~~~~~
```
Comment (SkafteNicki) [inline docs/source/references/modules.rst:466]:
```suggestion
PanopticQuality
~~~~~~~~~~~~~~~
```

File: docs/source/references/modules.rst [hunk 2]
```diff
@@ -462,6 +462,12 @@ MeanAveragePrecision
 .. autoclass:: torchmetrics.detection.mean_ap.MeanAveragePrecision
     :noindex:
 
+PanopticQuality
```
Comment (SkafteNicki) [inline docs/source/references/modules.rst:465]: similar instance should be added to `docs/source/references/functional.rst` for the functional counterpart

File: torchmetrics/detection/panoptic_quality.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+            target: ``torch.IntTensor`` ground truth of shape [height, width, 2] containing
+            the pair (category_id, instance_id) for each pixel of the image.
+            If the category_id refer to a stuff, the instance_id is ignored.
+
+        Raises:
+            ValueError:
+                If ``preds`` or ``target`` has wrong shape.
+            ValueError:
+                If ``preds`` containts
```
Comment (SkafteNicki) [inline torchmetrics/detection/panoptic_quality.py:118]: Something seems to be missing?

File: torchmetrics/functional/detection/panoptic_quality.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    return dict(zip(_totuple(unique_keys), unique_keys_area))
+
+
+def _is_set_int(value) -> bool:
+    """Check wheter value is a `Set[int]`"""
+    return isinstance(value, Set) and set(map(type, value)).issubset({int})
+
+
+def _validate_categories(things: Set[int], stuff: Set[int]):
+    if not _is_set_int(things):
```
Comment (SkafteNicki) [inline torchmetrics/functional/detection/panoptic_quality.py:45]:
```suggestion
def _validate_categories(things: Set[int], stuff: Set[int]):
    """Validate metric arguments `things` and `stuff`."""
    if not _is_set_int(things):
```

File: torchmetrics/functional/detection/panoptic_quality.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        raise ValueError("Expected argument `things` to be of type `Set[int]`")
+    if not _is_set_int(stuff):
+        raise ValueError("Expected argument `stuff` to be of type `Set[int]`")
+    if stuff & things:
+        raise ValueError("Expected arguments `things` and `stuffs` to have distinct keys.")
+
+
+def _validate_inputs(preds: torch.Tensor, target: torch.Tensor) -> None:
+    if not isinstance(preds, torch.Tensor):
```
Comment (SkafteNicki) [inline torchmetrics/functional/detection/panoptic_quality.py:54]:
```suggestion
def _validate_inputs(preds: torch.Tensor, target: torch.Tensor) -> None:
    """Validate predictions and target have the correct shape."""
    if not isinstance(preds, torch.Tensor):
```

File: torchmetrics/functional/detection/panoptic_quality.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        raise ValueError("Expected argument `preds` to be of type `torch.Tensor`")
+    if not isinstance(target, torch.Tensor):
+        raise ValueError("Expected argument `target` to be of type `torch.Tensor`")
+    if preds.shape != target.shape:
+        raise ValueError("Expected argument `preds` and `target` to have the same shape")
+    if preds.dim() != 3 or preds.shape[-1] != 2:
+        raise ValueError("Expected argument `preds` to have shape [height, width, 2]")
+
+
+def _get_void_color(things: Set[int], stuff: Set[int]):
```
Comment (SkafteNicki) [inline torchmetrics/functional/detection/panoptic_quality.py:64]: missing docsting

File: torchmetrics/functional/detection/panoptic_quality.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    unused_category_id = 1 + max([0] + list(things) + list(stuff))
+    return (unused_category_id, 0)
+
+
+def _get_category_id_to_continous_id(things: Set[int], stuff: Set[int]):
```
Comment (SkafteNicki) [inline torchmetrics/functional/detection/panoptic_quality.py:69]: missing docsting

File: torchmetrics/functional/detection/panoptic_quality.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    return (arr[..., None] == arr.new(values)).any(-1)
+
+
+def _prepocess_image(
+    things: Set[int],
+    stuff: Set[int],
+    img: torch.Tensor,
+    void_color: Tuple[int, int],
+    allow_unknown_category: bool,
+) -> torch.Tensor:
```
Comment (SkafteNicki) [inline torchmetrics/functional/detection/panoptic_quality.py:91]: missing docstring

File: torchmetrics/functional/detection/panoptic_quality.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+
+
+def _panoptic_quality_compute(
+    things: Set[int],
+    stuff: Set[int],
+    iou_sum: torch.Tensor,
+    true_positives: torch.Tensor,
+    false_positives: torch.Tensor,
+    false_negatives: torch.Tensor,
+) -> Dict:
```
Comment (SkafteNicki) [inline torchmetrics/functional/detection/panoptic_quality.py:184]: missing docstring

File: torchmetrics/functional/detection/panoptic_quality.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+    return metrics
+
+
+def panoptic_quality(
+    preds: torch.Tensor,
+    target: torch.Tensor,
+    things: Set[int],
+    stuff: Set[int],
+    allow_unknown_preds_category: bool = False,
+) -> Tensor:
```
Comment (SkafteNicki) [inline torchmetrics/functional/detection/panoptic_quality.py:224]: insert the docstring from the functional implementation here

File: torchmetrics/functional/detection/panoptic_quality.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+            sq=torch.mean(segmentation_quality[: len(things)]),
+            n=len(things),
+        ),
+        stuff=dict(
+            pq=torch.mean(panoptic_quality[len(things) :]),
+            rq=torch.mean(recognition_quality[len(things) :]),
+            sq=torch.mean(segmentation_quality[len(things) :]),
+            n=len(stuff),
+        ),
+    )
```
Comment (SkafteNicki) [inline torchmetrics/functional/detection/panoptic_quality.py:213]: I do not understand why a dict is constructed with the `all`, `things`, `stuff` keys initialized, but what is actually returned to the user is just: `results["all"]["pq"]`.
Seems it would be more efficient to just return `torch.mean(panoptic_quality)` directly

**Review by SkafteNicki — COMMENTED** [review]
few comments

**Comment by Borda** [comment]
@niberger, how is it going here? :rabbit:

**Comment by Borda** [comment]
@niberger how is it going here? :rabbit: 
we are considering a new feature release next week, do you think we can finish this addition by then so it will be included? :pray:

**Comment by Borda** [comment]
@niberger how is it going here? :rabbit:

**Comment by Borda** [comment]
@SkafteNicki I have checked the collisions, but do you think we could take it over and finish it? :otter:

File: src/torchmetrics/detection/panoptic_quality.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    _panoptic_quality_update,
+    _prepocess_image,
+    _validate_categories,
+    _validate_inputs,
+)
+from torchmetrics.metric import Metric
+
+
+class PanopticQuality(Metric):
+    r"""Computes the `Panoptic Quality`_ for panoptic segmentations. It is defined as:
```
Comment (Borda) [inline src/torchmetrics/detection/panoptic_quality.py:33]:
```suggestion
    r"""Compute the `Panoptic Quality`_ for panoptic segmentations.
```

File: src/torchmetrics/detection/panoptic_quality.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        >>> pred = torch.tensor([[[6, 0], [0, 0], [6, 0], [6, 0]],
+        ...                      [[0, 0], [0, 0], [6, 0], [0, 1]],
+        ...                      [[0, 0], [0, 0], [6, 0], [0, 1]],
+        ...                      [[0, 0], [7, 0], [6, 0], [1, 0]],
+        ...                      [[0, 0], [7, 0], [7, 0], [7, 0]]])
+        >>> target = torch.tensor([[[6, 0], [0, 1], [6, 0], [0, 1]],
+        ...                        [[0, 1], [0, 1], [6, 0], [0, 1]],
+        ...                        [[0, 1], [0, 1], [6, 0], [1, 0]],
+        ...                        [[0, 1], [7, 0], [1, 0], [1, 0]],
+        ...                        [[0, 1], [7, 0], [7, 0], [7, 0]]])
```
Comment (Borda) [inline src/torchmetrics/detection/panoptic_quality.py:64]:
```suggestion
        >>> from torch import tensor
        >>> pred = tensor([[[6, 0], [0, 0], [6, 0], [6, 0]],
        ...                [[0, 0], [0, 0], [6, 0], [0, 1]],
        ...                [[0, 0], [0, 0], [6, 0], [0, 1]],
        ...                [[0, 0], [7, 0], [6, 0], [1, 0]],
        ...                [[0, 0], [7, 0], [7, 0], [7, 0]]])
        >>> target = tensor([[[6, 0], [0, 1], [6, 0], [0, 1]],
        ...                  [[0, 1], [0, 1], [6, 0], [0, 1]],
        ...                  [[0, 1], [0, 1], [6, 0], [1, 0]],
        ...                  [[0, 1], [7, 0], [1, 0], [1, 0]],
        ...                  [[0, 1], [7, 0], [7, 0], [7, 0]]])
```

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+        nested_list:  the nested list to convert to a nested tuple
+
+    Returns:
+        a nested tuple with the same content.
+    """
+    """Construct a nested tuple from a nested list."""
+    return tuple(map(_nested_tuple, nested_list)) if isinstance(nested_list, list) else nested_list
+
+
+def _totuple(t: torch.Tensor) -> Tuple:
```
Comment (Borda) [inline src/torchmetrics/functional/detection/panoptic_quality.py:33]:
```suggestion
def _totuple(t: Tensor) -> Tuple:
```

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+    Args:
+        t: the tensor to convert
+
+    Returns:
+        a nested tuple with the same content.
+    """
+    return _nested_tuple(t.tolist())
+
+
+def _get_color_areas(img: torch.Tensor) -> Dict[Tuple, torch.Tensor]:
```
Comment (Borda) [inline src/torchmetrics/functional/detection/panoptic_quality.py:45]:
```suggestion
def _get_color_areas(img: Tensor) -> Dict[Tuple, Tensor]:
```

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+    return (arr[..., None] == arr.new(values)).any(-1)
+
+
+def _prepocess_image(
+    things: Set[int],
+    stuff: Set[int],
+    img: torch.Tensor,
+    void_color: Tuple[int, int],
+    allow_unknown_category: bool,
+) -> torch.Tensor:
```
Comment (Borda) [inline src/torchmetrics/functional/detection/panoptic_quality.py:159]:
```suggestion
def _prepocess_image(
    things: Set[int],
    stuff: Set[int],
    img: Tensor,
    void_color: Tuple[int, int],
    allow_unknown_category: bool,
) -> Tensor:
```

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+    img[~(things_pixels | stuff_pixels)] = img.new(void_color)
+    return img
+
+
+def _panoptic_quality_update(
+    flatten_preds: torch.Tensor,
+    flatten_target: torch.Tensor,
+    cat_id_to_continuous_id: Dict[int, int],
+    void_color: Tuple[int, int],
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
```
Comment (Borda) [inline src/torchmetrics/functional/detection/panoptic_quality.py:190]:
```suggestion
def _panoptic_quality_update(
    flatten_preds: Tensor,
    flatten_target: Tensor,
    cat_id_to_continuous_id: Dict[int, int],
    void_color: Tuple[int, int],
) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
```

**Review by Borda — APPROVED** [review]

**Comment by marcocaccin** [comment]
Hi @Borda, I just reached out on PTL's Slack to offer help if it's useful to have another pair of hands to bring this PR to the finish line ;) 
Quick question: `Metric.update()` is expected to work on inputs with a batch dimension, right? This implementation works for individual samples only.

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+        void_color: an additional, unused color
+
+    Returns:
+        - IOU Sum
+        - True positives
+        - False positives
+        - False negatives
+    """
+    r"""Returns stat scores (iou sum, true positives, false positives, false negatives) required to compute
+    accuracy."""
```
Comment (Borda) [inline src/torchmetrics/functional/detection/panoptic_quality.py:209]:
```suggestion
```

**Comment by justusschock** [comment]
Hey @marcocaccin 

yes, metrics should always work with batches. 
Help would be very welcome.

File: src/torchmetrics/detection/panoptic_quality.py [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    def __init__(
+        self,
+        things: Set[int],
+        stuffs: Set[int],
+        allow_unknown_preds_category: bool = False,
+        **kwargs: Any,
+    ):
+        super().__init__(**kwargs)
+
+        # todo
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/panoptic_quality.py:88]:
```suggestion
        # todo: better testing for correctness of metric
```

File: src/torchmetrics/detection/panoptic_quality.py [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+class PanopticQuality(Metric):
+    r"""Compute the `Panoptic Quality`_ for panoptic segmentations.
+
+    .. math::
+        PQ = \frac{IOU}{TP + 0.5 FP + 0.5 FN}
+
+    where IOU, TP, FP and FN are respectively the sum of the intersection over union for true positives, the number of
+    true postitives, false positives and false negatives. This metric is inspired by the PQ implementation of
+    panopticapi, a standard implementation for the PQ metric for object detection.
+
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/panoptic_quality.py:41]:
```suggestion

.. note:
    Metric is currently experimental

```

File: tests/unittests/detection/test_panoptic_quality.py [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+                [[0, 1], [7, 0], [7, 0], [1, 0], [1, 0]],
+                [[0, 1], [7, 0], [7, 0], [7, 0], [7, 0]],
+            ],
+        ]
+    ),
+)
+_args = {"things": {0, 1}, "stuffs": {6, 7}}
+
+
+def _compare_fn(preds, target) -> np.ndarray:
```
Comment (SkafteNicki) [inline tests/unittests/detection/test_panoptic_quality.py:70]:
```suggestion
def _compare_fn(preds, target) -> np.ndarray:
    """Reference implementation
    
    Improve this by calling https://github.com/cocodataset/panopticapi/blob/master/panopticapi/evaluation.py
    directly and compare at runtime on multiple examples
    """
```

**Review by SkafteNicki — APPROVED** [review]

**Comment by justusschock** [comment]
@marcocaccin we merged it for now. Feel free to do another PR to add batched-support and also maybe more tests :)

**Comment by ChristophReich1996** [comment]
Hi @niberger, I just wanted to ask if you have verified that your PQ implementation matches the results of the COCO PQ implementation? Thanks :)
